magento/magento2
# Prevent flushing same tags several times

### Description (*)
If a product is saved in admin - changed name, price, stock, weight, added category to product. Then after indexing, the same tags are flushed several times.
Here is an example of default Magento behaviour on cleaning cache tags during indexing (different arrays belong to different indexers cleaning the cache):
```
root@c119d6a06fcb:/var/www/m2clean$  bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1

Array
(
)


Array
(
)


Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [3] => cat_c_2
    [4] => cat_c_3
    [5] => cat_c_1
)

Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [3] => cat_c_2
    [4] => cat_c_3
    [5] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [3] => cat_c_2
    [4] => cat_c_3
    [5] => cat_c_1
)


Array
(
    [0] => cat_p_1
)

Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_c_2
    [5] => cat_c_1
    [6] => cat_c_2
    [7] => cat_c_3
    [8] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_c_2
    [5] => cat_c_1
    [6] => cat_c_2
    [7] => cat_c_3
    [8] => cat_c_1
)

Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_c_2
    [6] => cat_c_1
    [7] => cat_c_2
    [8] => cat_c_3
    [9] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_c_2
    [6] => cat_c_1
    [7] => cat_c_2
    [8] => cat_c_3
    [9] => cat_c_1
)

Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [8] => cat_c_2
    [9] => cat_c_3
    [10] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [8] => cat_c_2
    [9] => cat_c_3
    [10] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [8] => cat_c_2
    [9] => cat_c_3
    [10] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [8] => cat_c_2
    [9] => cat_c_3
    [10] => cat_c_1
)


Array
(
    [0] => cat_p_1
    [1] => cat_p_1
    [2] => cat_p_1
    [3] => cat_p_1
    [4] => cat_p_1
    [5] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [8] => cat_c_2
    [9] => cat_c_3
    [10] => cat_c_1
)
```

Here is the same output with the changes applied:
```
root@c119d6a06fcb:/var/www/m2clean$  bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1
Array
(
)


Array
(
)


Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [4] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [4] => cat_c_3
)


Array
(
    [0] => cat_p_1
    [1] => cat_c_2
    [2] => cat_c_1
    [4] => cat_c_3
)


Array
(
    [0] => cat_p_1
)

Array
(
    [0] => cat_p_1
    [4] => cat_c_2
    [5] => cat_c_1
    [7] => cat_c_3
)


Array
(
    [0] => cat_p_1
    [4] => cat_c_2
    [5] => cat_c_1
    [7] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [5] => cat_c_2
    [6] => cat_c_1
    [8] => cat_c_3
)


Array
(
    [0] => cat_p_1
    [5] => cat_c_2
    [6] => cat_c_1
    [8] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [9] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [9] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [9] => cat_c_3
)

Array
(
    [0] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [9] => cat_c_3
)


Array
(
    [0] => cat_p_1
    [6] => cat_c_2
    [7] => cat_c_1
    [9] => cat_c_3
)

```

### Related Pull Requests


### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. Observe the output of \Magento\Framework\Indexer\CacheContext::getIdentities during indexer run, same tags are flushed up to 10 times per one indexer instance
2. ...

### Questions or comments
Together with https://github.com/magento/magento2/pull/29699 it will probably reduce the amount of cache flushing significatly

### Contribution checklist (*)


### Resolved issues:
1. [x] resolves magento/magento2#29890: Prevent flushing same tags several times
